transport-cc parsing to be compliant with libwebrtc, make transport-cc reference time to be unsigned, therefore fix 357.

sarumjanuch opened the pull request:
Changes and tests to make sure mediasoup is parsing transport-cc RTCP feedback packages the same way as libwebrtc does. 
Metadata in commit was created by parsing buffers in libwebrtc project.
Also this PR includes changes https://github.com/versatica/mediasoup/pull/900 made by @penguinol. 
This PR makes sure that there could be no negative reference time (event though https://datatracker.ietf.org/doc/html/draft-holmer-rmcat-transport-wide-cc-extensions-01#section-3.1 mandates that reference time is 24 bit signed integer, libwebrtc is ignoring that, and with with PR we too), and also handles wraparound, to be consistent with latest libwebrtc code. Therefore it rolls back mediasoup changes in libwebrtc related to https://github.com/versatica/mediasoup/issues/357, and https://github.com/versatica/mediasoup/issues/357 should never happen again.

Commit:
Add test to check whether mediasoup parses transport-cc packets the same way as libwebrtc.

Commit:
Merge branch 'v3' of github.com:sarumjanuch/mediasoup into bwe-tests

ibc commented on worker/test/src/RTC/RTCP/TestFeedbackRtpTransport.cpp:
```diff
@@ -571,4 +571,223 @@ SCENARIO("RTCP Feeback RTP transport", "[parser][rtcp][feedback-rtp][transport]"
+                }
+        };
+
+        for (const auto &packetMeta: feedbackPacketsMeta) {
```
Please run `make lint` in worker/ directory. It should complain. `make format` will fix it.

Commit:
Fix Formatting.

Commit:
Reformat

ibc commented on worker/test/src/RTC/RTCP/TestFeedbackRtpTransport.cpp:
```diff
@@ -572,222 +572,163 @@ SCENARIO("RTCP Feeback RTP transport", "[parser][rtcp][feedback-rtp][transport]"
+		} FeedbackPacketsMeta;
+
+		// Metadata collected by parsing buffers with libwebrtc, buffers itself.
+		// were generated by chrome in direction of mediasoup.
```
```suggestion
		// Were generated by chrome in direction of mediasoup.
```

ibc committed:
Update worker/test/src/RTC/RTCP/TestFeedbackRtpTransport.cpp

ibc commented:
There are still lint errors: https://github.com/versatica/mediasoup/runs/8097882496?check_suite_focus=true

`make format` should fix them.

ibc commented:
Note that you can use `// clang-format off` and `on` to delimit C++ blocks for which lint should not be checked. Check `RequestChannel.cpp` file if you find it useful within this PR.

Commit:
Merge remote-tracking branch 'origin/bwe-tests' into bwe-tests

# Conflicts:
#	worker/test/src/RTC/RTCP/TestFeedbackRtpTransport.cpp

Commit:
Get rid of negative reference_time, make reference timestamp to be represented as in libwebrtc.

Commit:
Fix tests.
Adopt changes to GetBaseDelta from https://github.com/versatica/mediasoup/pull/900
Add tests for GetBaseDelta Wraparound.

Commit:
Fix tests.
Adopt changes to GetBaseDelta from https://github.com/versatica/mediasoup/pull/900
Add tests for GetBaseDelta Wraparound.

Commit:
Revert "Fix tests."

This reverts commit b4b0ac8758ded1b6099b40ccb585d5a3ad4a8ad0.

Commit:
Fix tests.
Adopt changes to GetBaseDelta from https://github.com/versatica/mediasoup/pull/900
Add tests for GetBaseDelta Wraparound.

ibc commented on worker/include/RTC/RTCP/FeedbackRtpTransport.hpp:
```diff
@@ -42,6 +42,8 @@ namespace RTC
 		{
 		public:
+			static constexpr int64_t kBaseTimeTick   = 64;
+			static constexpr int64_t kTimeWrapPeriod = kBaseTimeTick * (1ll << 24);
```
Blank line after this, please?

jmillan replied:
Also please, remove the "k" prefix from constant name as it's not mediasoup style but libwebrtc.

ibc replied:
And should it start with upper letter? I cannot remember.

sarumjanuch replied:
Made without k, and starting with a capital letter.

ibc commented on worker/include/RTC/RTCP/FeedbackRtpTransport.hpp:
```diff
@@ -237,9 +239,29 @@ namespace RTC
 			{
 				return this->referenceTime;
 			}
+			// We only use this for testing purpose
```
Dot at the end of the comment, please 😊

ibc commented on worker/include/RTC/RTCP/FeedbackRtpTransport.hpp:
```diff
@@ -290,7 +312,7 @@ namespace RTC
 		private:
 			uint16_t baseSequenceNumber{ 0u };
-			int32_t referenceTime{ 0 };
+			uint32_t referenceTime{ 0 };
```
Are we sure of this? Reference time is supposed to be a 24bits signed value.

sarumjanuch replied:
Yes, the whole point of this commit is to not follow that rule, as libwebrtc doesn't care about sign for a reference_time already. https://webrtc.googlesource.com/src/+/refs/heads/main/modules/rtp_rtcp/source/rtcp_packet/transport_feedback.h#170

Commit:
Cosmetics.

ibc commented on worker/include/RTC/RTCP/FeedbackRtpTransport.hpp:
```diff
@@ -237,9 +241,29 @@ namespace RTC
+			int64_t GetBaseDelta(const int64_t previousTimestampMs) const
+			{
+				int64_t delta = GetReferenceTimestamp() - previousTimestampMs;
+				// Compensate for wrap around.
```
Add a blank line before the comment, please.

ibc reviewed (approved):
Nice. Just a cosmetic change requested. Also, if this PR fixes an existing issue in GH (I think there is a reported crash, am I right?) reference it in the PR description so it will be automatically closed. Tell me when those things are done and let's merge then.

Commit:
Cosmetics.

sarumjanuch commented:
> Nice. Just a cosmetic change requested. Also, if this PR fixes an existing issue in GH (I think there is a reported crash, am I right?) reference it in the PR description so it will be automatically closed. Tell me when those things are done and let's merge then.
@ibc Seems like done, please run tests, because I am not sure that clang formatter is not playing games with me again.

ibc commented:
Running!

ibc commented:
Please re-request review once tests pass.

ibc commented:
> Nice. Just a cosmetic change requested. Also, if this PR fixes an existing issue in GH (I think there is a reported crash, am I right?) reference it in the PR description so it will be automatically closed. Tell me when those things are done and let's merge then.

Will merge. Can you check this please? I think there is an open issue or PR (here in GH) that aims to fix an issue that may be fixed by this PR, but maybe I'm wrong.

ibc commented:
Oh, I request a change: please add a item in Changelog.md following the style of other items.

sarumjanuch commented:
> Will merge. Can you check this please? I think there is an open issue or PR (here in GH) that aims to fix an issue that may be fixed by this PR, but maybe I'm wrong.

Yes, i did that, also added in title.

sarumjanuch commented:
> Oh, I request a change: please add a item in Changelog.md following the style of other items.

doing right now

Commit:
Update CHANGELOG.md
